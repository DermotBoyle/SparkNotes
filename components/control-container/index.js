export { ControlContainer } from './control-container.jsx'

export { SideMenu } from './side-menu.jsx'

export { ControlMenu } from './control-menu.jsx'
