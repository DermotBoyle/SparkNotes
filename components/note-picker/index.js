export { NotePicker } from './note-picker'
