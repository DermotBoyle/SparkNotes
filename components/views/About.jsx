import React from 'react'
import { useEffect, useState } from 'react'

export const About = ({ transitionStyles, defaultStyle, transitionState }) => {
  const [on, setOn] = useState('entering')

  useEffect(() => {
    setOn(transitionState)
  }, [])

  return (
    <section
      style={{
        ...defaultStyle,
        ...transitionStyles[on]
      }}
    >
      <p>This is an about App I built with NextJS and MongoDB</p>
    </section>
  )
}
