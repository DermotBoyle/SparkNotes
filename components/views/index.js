export { About } from './About'

export { Find } from './Find'

export { ViewAll } from './ViewAll'

export { Home } from './Home'

export { CreateNoteForm } from './CreateNoteForm'

export { EditNoteForm } from './EditNoteForm'
