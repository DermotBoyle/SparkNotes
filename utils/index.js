export * from './CONSTANTS'

export * from './helpers'

export { transitionManager } from './transition-manager/transition-manager'
