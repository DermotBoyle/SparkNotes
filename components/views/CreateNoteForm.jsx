import { SaveButton } from 'components/save-button'
import { useReducer } from 'react'
import router from 'next/router'

import styles from 'styles/create-note-form.module.scss'

import moment from 'moment'

import { Routes } from 'utils'

export const CreateNoteForm = ({
  transitionStyles,
  defaultStyle,
  transitionState
}) => {
  const initialFormValues = {
    subject: '',
    keywords: '',
    content: '',
    date: ''
  }

  const [formValues, setFormValues] = useReducer(
    (curVals, newVals) => ({ ...curVals, ...newVals }),
    initialFormValues
  )

  const { subject, keywords, content } = formValues

  const handleFormChange = ({ target }) => {
    const { name, value } = target
    setFormValues({ [name]: value })
  }

  const handleSaveClick = async () => {
    formValues.date = moment(new Date()).format('YYYY-MM-DD-T00:00:00.000Z')
    updateDB(formValues)
  }

  const updateDB = async formValues => {
    try {
      await fetch('api/note', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(formValues)
      }).then(({ statusText }) => {
        if (statusText === 'Created') {
          setFormValues(initialFormValues)
          if (res.status === 201) router.push(Routes.BASE + Routes.VIEW_ALL)
        }
      })
    } catch (error) {
      console.error(error)
    }
  }

  return (
    <section
      className={styles['form-container']}
      style={{
        ...defaultStyle,
        ...transitionStyles[transitionState]
      }}
    >
      <form>
        <div className={styles['keywords-container']}>
          <div>
            <label>Subject</label>
            <input
              type='text'
              name='subject'
              value={subject}
              onChange={handleFormChange}
            />
          </div>
          <div>
            <label>Keywords</label>
            <input
              type='text'
              name='keywords'
              value={keywords}
              onChange={handleFormChange}
            />
          </div>
        </div>
        <div className={styles['main-note']}>
          <label>Write your note here</label>
          <textarea
            type='text'
            name='content'
            value={content}
            onChange={handleFormChange}
          />
        </div>
      </form>
      <SaveButton formValues={formValues} handleSaveClick={handleSaveClick} />
    </section>
  )
}
