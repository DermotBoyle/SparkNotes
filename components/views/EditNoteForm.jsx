import { SaveButton } from 'components/save-button'
import { useReducer } from 'react'

import styles from 'styles/create-note-form.module.scss'

import { useAppContext } from 'context/global-state'
import Router from 'next/router'
import { Routes, Methods } from 'utils'

import moment from 'moment'
import { v4 as uuidv4 } from 'uuid'

export const EditNoteForm = () => {
  const {
    state: {
      subject: initialSubject,
      keywords: initialKeywords,
      content: initialContent,
      _id
    },
    dispatch
  } = useAppContext()

  const [formValues, setFormValues] = useReducer(
    (curVals, newVals) => ({ ...curVals, ...newVals }),
    {
      content: initialContent,
      keywords: initialKeywords,
      subject: initialSubject,
      currentKeyword,
      _id
    }
  )

  const { subject, keywords, content, currentKeyword } = formValues

  const handleFormChange = ({ target }) => {
    const { name, value } = target
    setFormValues({ [name]: value })
  }

  const formatKeywords = ({ keyCode, target: { value } }) => {
    if (keyCode === 32 || keyCode === 13) {
      setFormValues({ keywords: [...keywords, value], currentKeyword: '' })
    }
  }

  const handleSaveClick = async () => {
    formValues.date = moment.utc().format()
    updateDB(formValues)
  }

  const handleRemoveKeyword = value => {
    const updatedKeywords = formValues.keywords.filter(word => word !== value)
    setFormValues({ keywords: updatedKeywords })
  }

  /* const updateDB = async formValues => {
    try {
      await fetch(`api/note/${formValues._id}`, {
        method: Methods.PUT,
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(formValues)
      }).then(res => {
        if (res.status === 201) Router.push(Routes.BASE + Routes.VIEW_ALL)
      })
    } catch (error) {
      alert(error)
    }
  }*/

  const updateDB = ( formValues ) => {
    dispatch({
      type: 'updateAtId',
      payload: {currentNote: formValues}
    })
    Router.push(Routes.BASE + Routes.VIEW_ALL)
  }

  return (
    <section
      className={styles['form-container']}
    >
      <form>
        <div className={styles['inputs-container']}>
          <div className={styles['subject-input']}>
            <label>Subject</label>
            <input
              type='text'
              name='subject'
              value={subject}
              onChange={handleFormChange}
            />
          </div>
          <div className={styles['keywords-input']}>
            <label>Keywords</label>
            <p className={styles['keywords-saved']}>{keywords?.length}/ 5</p>
            <input
              disabled={keywords?.length === 5}
              type='text'
              name='currentKeyword'
              value={currentKeyword}
              onChange={handleFormChange}
              onKeyDown={formatKeywords}
            />
            <div className={styles['keyword-chips-container']}>
              {!!formValues?.keywords?.length
                ? keywords?.map(word => (
                    <p onClick={() => handleRemoveKeyword(word)} key={uuidv4()}>
                      {word} X
                    </p>
                  ))
                : null}
            </div>
          </div>
        </div>
        <div className={styles['main-note']}>
          <label>Write your note here</label>
          <textarea
            type='text'
            name='content'
            value={content}
            onChange={handleFormChange}
          />
        </div>
      </form>
      <SaveButton formValues={formValues} handleSaveClick={handleSaveClick} />
    </section>
  )
}
