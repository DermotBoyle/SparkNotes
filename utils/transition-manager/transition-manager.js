import React from 'react'
import {
  Home,
  ViewAll,
  About,
  Find,
  CreateNoteForm,
  EditNoteForm
} from 'components/views'
import { Routes } from 'utils'

export const transitionManager = () => {
  const View = {
    [Routes.HOME]: (
      <Home/>
    ),
    [Routes.VIEW_ALL]: (
      <ViewAll />
    ),
    [Routes.FIND]: (
      <Find/>
    ),
    [Routes.ABOUT]: (
      <About/>
    ),
    [Routes.CREATE]: (
      <CreateNoteForm/>
    ),
    [Routes.EDIT]: (
      <EditNoteForm/>
    )
  }

  return function ComponentToRender ({ currentURL }) {
    return View[currentURL]
  }
}

