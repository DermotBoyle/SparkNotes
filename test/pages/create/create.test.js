/* eslint-disable react/display-name */
import * as React from 'react'
import { CreateNoteForm } from 'components/views'
import { transitionStyles } from 'utils/transition-manager/transition-manager'
import { fireEvent, render, screen } from '@testing-library/react'
import { WrapperProviderComp, allNotes } from "../../test.utils"



// eslint-disable-next-line @next/next/no-img-element
jest.mock('next/image', () => ({ src = "", alt }) => <img src={src} alt={alt} />)
jest.mock('next/router', ()=> ({push: jest.fn()}))


describe('<CreateNoteForm/>', () => {
 
    it('should call the dispatch fn',async () => {
        const mockDispatch = jest.fn()
        const mockState = { currentURL: '/create', savedNotes: allNotes }

        const { baseElement } = render(
          WrapperProviderComp(
            mockDispatch,
            mockState
          )(<CreateNoteForm transitionStyles={transitionStyles} />)
        )
    
        const subjectInput = baseElement.querySelector(`input[name="subject"]`)
        fireEvent.change(subjectInput, { target: { value: "TEST SUBJECT" } })
        fireEvent.submit(subjectInput)

        const keywordsInput = baseElement.querySelector(`input[name="currentKeyword"]`)
        fireEvent.change(keywordsInput, { target: { value: "TEST KEYWORD"}})
        fireEvent.submit(keywordsInput)

        const contentInput = baseElement.querySelector(`textarea[name="content"]`)
        fireEvent.change(contentInput, { target: { value: "TEST CONTENT"}})
        fireEvent.submit(contentInput)

        const button = screen.getByRole("button")
        fireEvent.click(button)

        expect(baseElement).toMatchSnapshot()
        expect(mockDispatch).toHaveBeenCalled()
   
      })

})