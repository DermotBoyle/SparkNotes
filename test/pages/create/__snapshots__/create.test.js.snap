// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateNoteForm/> should call the dispatch fn 1`] = `
<body>
  <div>
    <section
      class="form-container"
    >
      <form>
        <div
          class="inputs-container"
        >
          <div
            class="subject-input"
          >
            <label>
              Subject
            </label>
            <input
              name="subject"
              type="text"
              value="TEST SUBJECT"
            />
          </div>
          <div
            class="keywords-input"
          >
            <label>
              Keywords
            </label>
            <p
              class="keywords-saved"
            >
              0
              / 5
            </p>
            <input
              name="currentKeyword"
              type="text"
              value="TEST KEYWORD"
            />
            <div
              class="keyword-chips-container"
            />
          </div>
        </div>
        <div
          class="main-note"
        >
          <label>
            Write your note here
          </label>
          <textarea
            name="content"
            type="text"
          >
            TEST CONTENT
          </textarea>
        </div>
      </form>
      <button
        aria-label="save new note"
        class="save-button"
        data-for="save-button"
        data-tip="true"
      >
        <img
          alt="tick symbol"
          src="[object Object]"
        />
      </button>
    </section>
  </div>
</body>
`;
